### PR TITLE
Add rm --filter option

### PR DESCRIPTION
**Pull request by sstosh**
--filter : remove the filtered container.

Signed-off-by: Toshiki Sonoda <sonoda.toshiki@fujitsu.com>

#### Does this PR introduce a user-facing change?

```release-note
podman rm now supports the --filter option.
```

### CODE REVIEW

File: cmd/podman/containers/rm.go [hunk 1]
```diff
@@ -69,9 +71,13 @@ func rmFlags(cmd *cobra.Command) {
 	flags.BoolVarP(&rmOptions.Volumes, "volumes", "v", false, "Remove anonymous volumes associated with the container")
 
 	cidfileFlagName := "cidfile"
-	flags.StringArrayVar(&cidFiles, cidfileFlagName, nil, "Read the container ID from the file")
+	flags.StringArrayVar(&rmCidFiles, cidfileFlagName, nil, "Read the container ID from the file")
```
Comment (vrothberg) [inline cmd/podman/containers/rm.go:74]: `killCidFiles` is initialized with `[]string{}` not `nil`.  Is there any noticeable difference?

File: pkg/domain/infra/abi/containers.go [hunk 2]
```diff
@@ -380,17 +380,27 @@ func (ic *ContainerEngine) ContainerRm(ctx context.Context, namesOrIds []string,
...
-	ctrs, err := getContainersByContext(options.All, options.Latest, names, ic.Libpod)
+	ctrs, rawInputs, err := getContainersAndInputByContext(options.All, options.Latest, names, options.Filters, ic.Libpod)
 	if err != nil && !(options.Ignore && errors.Is(err, define.ErrNoSuchCtr)) {
 		// Failed to get containers. If force is specified, get the containers ID
 		// and evict them
 		if !options.Force {
 			return nil, err
 		}
 
+		ctrMap := map[string]string{}
```
Comment (vrothberg) [inline pkg/domain/infra/abi/containers.go:391]: The name `ctrMap` is very generic. Can we give it a more expressive name, for instance, `idToRawInput`?

File: pkg/domain/infra/tunnel/containers.go [hunk 3]
```diff
@@ -208,11 +208,18 @@ func (ic *ContainerEngine) ContainerRm(ctx context.Context, namesOrIds []string,
 
 	toRemove := []string{}
 	alreadyRemoved := make(map[string]bool) // Avoids trying to remove already removed containers
-	if opts.All {
-		ctrs, err := getContainersByContext(ic.ClientCtx, opts.All, opts.Ignore, nil)
+	ctrMap := map[string]string{}
```
Comment (vrothberg) [inline pkg/domain/infra/tunnel/containers.go:211]: Same here.

**Review by vrothberg — COMMENTED** [review]
Thanks for contributing, @sstosh!

Can you elaborate why we need to preserve the raw inputs?

File: cmd/podman/containers/rm.go (hunk 1 above)
Comment (sstosh) [reply cmd/podman/containers/rm.go:74]: I think, there is not functional difference between `[]string{}` and `nil`.
I'll unify the arguments to nil.

**Comment by sstosh** [comment]
> Thanks for contributing, @sstosh!
> 
> Can you elaborate why we need to preserve the raw inputs?

`rawInputs` is used to display the target container ID.
I don't think there is any other reason.
I think there is a smarter way to cording.

**Comment by vrothberg** [comment]
> rawInputs is used to display the target container's ID.

I'd feel more comfortable if we could discuss and tackle it in a separate PR.  Or am I missing the point and it's critical for the filters?

**Comment by sstosh** [comment]
> I'd feel more comfortable if we could discuss and tackle it in a separate PR. Or am I missing the point and it's critical for the filters?

`ctrMap` is used not only `rm` but also `stop`, `kill`, `pause`, `unpause`,
so we may need a lot of fixes.
If we need to fix it, we should open a new PR.

**Comment by vrothberg** [comment]
Thanks, @sstosh.  I assume it's to make sure that `stop/foo foo` will print "foo" instead of the ID.  I am OK putting it in this PR.

File: test/e2e/rm_test.go [hunk 4]
```diff
@@ -282,4 +283,45 @@ var _ = Describe("Podman rm", func() {
...
+		cid3 := session1.OutputToString()
+		shortCid3 := cid3[0:5]
+
+		session1 = podmanTest.Podman([]string{"start", "--all"})
+		session1.WaitWithDefaultTimeout()
+		Expect(session1).Should(Exit(0))
+
+		session1 = podmanTest.Podman([]string{"rm", cid1, "-f", "--filter", "status=running"})
+		session1.WaitWithDefaultTimeout()
+		Expect(session1).Should(Exit(125))
```
Comment (vrothberg) [inline test/e2e/rm_test.go:310]: Please check the output of `rm` here and at the other places.

**Review by vrothberg — COMMENTED** [review]
Other than the last it, LGTM

@containers/podman-maintainers PTAL

**Pull request merged**